lib/Conversion/CalyxToHW/CalyxToHW.cpp

```diff
@@ -13,45 +13,369 @@
-    for (auto portInfo : enumerate(portInfos)) {
-      hw::PortInfo hwPortInfo;
-      hwPortInfo.direction = direction;
-      hwPortInfo.type = portInfo.value().type;
-      hwPortInfo.argNum = portInfo.index();
-      hwPortInfo.name = portInfo.value().name;
-      hwPortInfos.push_back(hwPortInfo);
+/// ConversionPatterns.
+
+struct ConvertProgramOp : public OpConversionPattern<ProgramOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ProgramOp program, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    ModuleOp mod = program->getParentOfType<ModuleOp>();
+    rewriter.inlineRegionBefore(program.body(), &mod.getBodyRegion().front());
+    rewriter.eraseBlock(&mod.getBodyRegion().back());
+    return success();
+  }
+};
+
+struct ConvertComponentOp : public OpConversionPattern<ComponentOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ComponentOp component, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    SmallVector<hw::PortInfo> hwInputInfo;
+    for (auto portInfo : component.getPortInfo())
+      hwInputInfo.push_back(
+          {portInfo.name, hwDirection(portInfo.direction), portInfo.type});
+
+    auto hwMod = rewriter.create<HWModuleOp>(
+        component.getLoc(), component.getNameAttr(), hwInputInfo);
+
+    rewriter.eraseOp(hwMod.getBodyBlock()->getTerminator());
+    ConversionPatternRewriter::InsertionGuard g(rewriter);
+    rewriter.setInsertionPointToEnd(hwMod.getBodyBlock());
+
+    SmallVector<Value> argValues;
+    SmallVector<Value> outputWires;
+    size_t portIdx = 0;
+    for (auto portInfo : component.getPortInfo()) {
+      switch (portInfo.direction) {
+      case calyx::Direction::Input:
+        assert(hwMod.getArgument(portIdx).getType() == portInfo.type);
+        argValues.push_back(hwMod.getArgument(portIdx));
+        break;
+      case calyx::Direction::Output:
+        auto wire = rewriter.create<sv::WireOp>(component.getLoc(),
+                                                portInfo.type, portInfo.name);
+        auto wireRead =
+            rewriter.create<sv::ReadInOutOp>(component.getLoc(), wire);
+        argValues.push_back(wireRead);
+        outputWires.push_back(wireRead);
+        break;
+      }
+      ++portIdx;
     }
-  };
 
-  addHWPortInfo(op.getInputPortInfo(), hw::PortDirection::INPUT);
-  addHWPortInfo(op.getOutputPortInfo(), hw::PortDirection::OUTPUT);
+    rewriter.mergeBlocks(component.getBody(), hwMod.getBodyBlock(), argValues);
+    rewriter.create<OutputOp>(component.getLoc(), outputWires);
+    rewriter.eraseOp(component);
 
-  return builder.create<hw::HWModuleExternOp>(
-      op->getLoc(), builder.getStringAttr(op.getName()), hwPortInfos);
-}
+    return success();
+  }
+
+private:
+  hw::PortDirection hwDirection(calyx::Direction dir) const {
```

Comment:
Should this be moved to CalyxOps.h?

Reply:
I'd just as soon avoid CalyxOps.h depending on HWOps.h. I sort of think the only place we'd need this mapping is here, while converting Calyx->HW. If it's useful elsewhere, I can factor it out easily enough.